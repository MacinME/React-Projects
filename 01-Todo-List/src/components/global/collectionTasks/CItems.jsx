import { useState } from 'react';
import Close from '../assets/close.svg';

export const CItems = ({ 
    id, 
    taskName, 
    date,
    status,
    onDeleteDataFromLocalStorage,
    tasks,
    onUpdateLocalStorage
  }) => {

    const [ischecked, setIschecked] = useState( status );

    const onSuccessTask = (evt) => {

      setIschecked(evt.target.checked);

      const updatedTasks = tasks.map( task => {
        if(task.id === id){
          task.status = evt.target.checked;
        }
        return task;
      });
    
      onUpdateLocalStorage(updatedTasks);
    }

  return (
    <div className="bg-navbar_color rounded-xl w-full py-1 px-2">
        <div className="flex justify-between">
          <div className='flex gap-2'>
            <input 
                  type="checkbox"
                  onChange={ onSuccessTask } 
                  checked={ ischecked }
                  className="accent-pink-500 rounded-full"
              />
              <div>
                  <p className={`text-gray-100 text-sm lg:text-base ${ ischecked && 'line-through' }`}>{ taskName }</p>
                  <p className="text-gray-400 text-sm">Date: <span className="text-button_color">{ date }</span></p>
              </div>
          </div>

          <button
            onClick={ () => onDeleteDataFromLocalStorage(id) }
            className='text-white'
          >
            <img 
              src={ Close } 
              alt="Icon"
              className='w-4 md:w-6' 
            />
          </button>
        </div>
    </div>
  )
}
