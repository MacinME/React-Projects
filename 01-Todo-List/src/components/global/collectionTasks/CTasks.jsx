import { useEffect } from 'react';
import PropTypes from 'prop-types';
import Plus from '../../../assets/smallPlus.svg'
import { useForm } from '../../../hooks/useForm';
import ALeft from '../assets/aLeft.svg';
import { CItems } from './CItems';
import { useLocalStorage } from '../../../hooks/useLocalStorage';

export const CTasks = ({ state, setState }) => {
    const { collection: { title, id } } = state;

    const { tasks, setTasks, onGetDataById, onDeleteDataFromLocalStorage, onUpdateLocalStorage } = useLocalStorage(id);
    const { simpleForm, taskName, onInputChange, onResetForm } = useForm({
        id: null,
        taskName: '',
        date: new Date().toLocaleDateString(),
        status: false
    });

    const onAddTask = (e) => {
        e.preventDefault();
        if( taskName.trim().length < 2 ) return;    
        const { collectionTasks } = onGetDataById(id);
        const newValues = [...collectionTasks, {
            ...simpleForm,
            id: Date.now()
        }];

        onUpdateLocalStorage( newValues );
        onResetForm();
    }

    useEffect(() => {
        const data = onGetDataById(id);
        if(data){
            setTasks( data.collectionTasks );
        }
    }, [id])

  return (
    <div className="w-full flex flex-col gap-5 px-4 lg:w-3/6 lg:mx-auto pt-5 lg:pt-20">
        <div className="w-full bg-background_color flex flex-col gap-3 fixed top-0 left-0 py-4 px-3 lg:w-3/6 lg:top-12 lg:left-1/4">
            <div className="flex items-center gap-4 rounded-xl">
                <button
                    data-testid="back"
                    onClick={ () => setState({ status: false, collection: [] }) }
                    className='w-10 h-10 bg-navbar_color flex items-center justify-center rounded-xl py-1'
                >
                    <img 
                        src={ ALeft } 
                        alt="Icon" 
                        className='w-6'
                    />
                </button>
                <p className="text-white texxt-2xl font-bold lg:text-2xl">
                    { title }
                </p>
            </div>
            <form
                aria-label="form"
                onSubmit={ onAddTask }
                className="w-full flex items-center gap-3 bg-background_color px-2 rounded-xl border border-gray-700"
            >
                <button
                    className="bg-button_color w-8 h-7 text-dark-900 flex items-center justify-center rounded-xl cursor-pointer"
                >
                    <img 
                        src={ Plus } 
                        alt="Icon" 
                        className='w-6 '
                    />
                </button>
                <input 
                    type="text" 
                    placeholder="Add Task"
                    id='taskName'
                    onChange={ onInputChange }
                    name='taskName'
                    value={ taskName }
                    className="w-full bg-transparent py-2 text-gray-400 text-sm outline-none focus:text-white"
                />
            </form>
        </div>

        <div className='flex flex-col gap-3 pt-28 pb-24'>
            {
                tasks.map( task => (
                    <CItems
                        key={ task.id }
                        {...task}
                        tasks={ tasks } 
                        onUpdateLocalStorage={ onUpdateLocalStorage }
                        onDeleteDataFromLocalStorage={ onDeleteDataFromLocalStorage }
                    />
                ))
            }
        </div>
    </div>
  )
}

CTasks.propTypes = {
    state: PropTypes.object.isRequired,
    setState: PropTypes.func.isRequired
}