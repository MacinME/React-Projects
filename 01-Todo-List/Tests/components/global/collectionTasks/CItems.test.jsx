import { render } from '@testing-library/react';
import { CItems } from '../../../../src/components/global/collectionTasks/CItems';

import mockedSVGContent from '../../../__mocks__/imageMocked';

jest.mock('../../../../src/components/global/assets/close.svg', () => ({
    ReactComponent: () => <div data-testid="mocked-svg" dangerouslySetInnerHTML={{ __html: mockedSVGContent }}  />
}));

describe('Test on <CItems /> component', () => {

    // Arrange
    const task = {
        id: 12345,
        taskName: 'Learning Testing with React Library',
        date: new Date().toLocaleDateString(),
        status: false,
        onDeleteDataFromLocalStorage: jest.fn(),
        tasks: [],
        onUpdateLocalStorage: jest.fn()
    }
    
    test('should have the task an date strings ', () => { 

        // Act
        const { container } = render(
            <CItems 
                {...task}
            />
        )

        // Assert
        expect( container.getElementsByTagName('p').length ).toBeGreaterThanOrEqual( 2 );
        const [ taskName, taskDate ] = container.getElementsByTagName('p');
        expect( taskName.innerHTML ).toBe( task.taskName );
        expect( taskDate.textContent ).toContain( task.date );
     })

     test('should match with the snapshot', () => {

        const { container} = render(
            <CItems 
                {...task}
            />
        )

        expect( container ).toMatchSnapshot();

     })
});