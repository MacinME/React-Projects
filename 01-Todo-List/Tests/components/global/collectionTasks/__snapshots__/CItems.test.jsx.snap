// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <CItems /> component should match with the snapshot 1`] = `
<div>
  <div
    class="bg-navbar_color rounded-xl w-full py-1 px-2"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="flex gap-2"
      >
        <input
          class="accent-pink-500 rounded-full"
          id="12345"
          type="checkbox"
        />
        <div>
          <p
            class="text-gray-100 text-sm lg:text-base false"
          >
            Learning Testing with React Library
          </p>
          <p
            class="text-gray-400 text-sm"
          >
            Date: 
            <span
              class="text-button_color"
            >
              7/19/2023
            </span>
          </p>
        </div>
      </div>
      <button
        class="text-white"
      >
        <img
          alt="Icon"
          class="w-4 md:w-6"
          src="[object Object]"
        />
      </button>
    </div>
  </div>
</div>
`;
